src/program.cpp

```diff
@@ -828,77 +829,124 @@ std::string cvk_program::prepare_build_options(const cvk_device* device) const {
         }
-    }
-#endif
 
-    // Prepare build options
-    auto build_options = prepare_build_options(device);
-    if (build_from_il) {
-        build_options += " -x ir ";
-    }
+        cmd += clspv_input_file;
+        cmd += " ";
 
-    // Save headers
-    if (use_tmp_folder) {
-        build_options += "-I" + tmp_folder;
-        for (cl_uint i = 0; i < m_num_input_programs; i++) {
-            std::string fname{tmp_folder + "/" + m_header_include_names[i]};
-            if (!save_string_to_file(fname, m_input_programs[i]->source())) {
-                cvk_error_fn("Couldn't save header to file!");
+    } else if (m_operation == build_operation::link) {
+        for (auto input_program : m_input_programs) {
+            if (input_program->m_binary_type !=
+                    CL_PROGRAM_BINARY_TYPE_COMPILED_OBJECT &&
+                input_program->m_binary_type !=
+                    CL_PROGRAM_BINARY_TYPE_LIBRARY) {
                 return CL_BUILD_ERROR;
             }
+            std::string input_file = clspv_input_file + "_" +
+                                     std::to_string((uintptr_t)input_program) +
+                                     ".bc";
+            if (!save_il_to_file(input_file, input_program->m_ir)) {
+                cvk_error_fn("Couldn't save source to file!");
+                return CL_BUILD_ERROR;
+            }
+            cmd += input_file;
+            cmd += " ";
+        }
+    } else {
+        if (m_source.empty() && !m_ir.empty()) {
+            clspv_input_file += ".bc";
+            if (!save_il_to_file(clspv_input_file, m_ir)) {
+                cvk_error_fn("Couldn't save source to file!");
+                return CL_BUILD_ERROR;
+            }
+            cmd += clspv_input_file;
+            cmd += " ";
+        } else {
+            clspv_input_file += ".cl";
+            if (!save_string_to_file(clspv_input_file, m_source)) {
+                cvk_error_fn("Couldn't save source to file!");
+                return CL_BUILD_ERROR;
+            }
+            cmd += clspv_input_file;
+            cmd += " ";
         }
     }
 
-    // Select operation
-    // TODO support building a library with clBuildProgram
-    if (m_operation == build_operation::compile) {
-        m_binary_type = CL_PROGRAM_BINARY_TYPE_COMPILED_OBJECT;
+    std::string spirv_file{tmp_folder + "/compiled.spv"};
+    cmd += build_options;
+    cmd += " -o ";
+    cmd += spirv_file;
+
+    // Call clspv
+    int status = cvk_exec(cmd, &m_build_log);
+    if (status != 0) {
+        cvk_error_fn("failed to compile the program");
+        return CL_BUILD_ERROR;
+    }
+
+    if (build_to_ir) {
+        std::ifstream stream(spirv_file, std::ios::in | std::ios::binary);
+        m_ir.assign((std::istreambuf_iterator<char>(stream)),
+                    std::istreambuf_iterator<char>());
+        cvk_info("Loaded IR binary from \"%s\", size = %zu", spirv_file.c_str(),
+                 spirv_file.size());
```

Comment:
This is reporting the binary's size as the length of the name of the file it was loaded from :)